### PR TITLE
Adding Components

### PR DESCRIPTION
**Pull request by Snayak7643**

### CODE REVIEW

File: src/type.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
+export type productType = {
```
Comment (Ishaan-flyfin) [inline src/type.ts:1]: nice!

File: src/App.tsx [hunk 2]
```diff
@@ -1,7 +1,40 @@
...
+  useEffect(() => {
+    let total = 0;
+    const calculateTotal = () => {
+      productsData.forEach((product) => {
+        total = total + product.price;
+      });
+    };
+    calculateTotal();
+    dispatch({
+      type: ACTIONS.SET_INITIAL,
```
Comment (Ishaan-flyfin) [inline src/App.tsx]: we can calculate amount and total within SET_INITIAL reducer

File: src/Constants/InitialCartState.tsx [hunk 3]
```diff
@@ -0,0 +1,8 @@
+import productsData from "../data";
+import { State } from "../type";
+
+export const initialCartState: State = {
+  productsData,
+  amount: 4,
+  total: 2599.96,
```
Comment (Ishaan-flyfin) [inline src/Constants/InitialCartState.tsx]: we can keep it [],0,0

File: src/Reducers/functions.tsx [hunk 4]
```diff
@@ -0,0 +1,30 @@
+import { State } from "../type";
+
+export const increaseAmount = (state: State, price: number) => {
+  return {
+    ...state,
+    amount: state.amount + 1,
+    total: Math.round((state.total + price) * 100) / 100,
+  };
+};
```
Comment (Ishaan-flyfin) [inline src/Reducers/functions.tsx]: export const increaseAmount = (id:number, amount: number, price: number) => {
  return {
        type: ACTIONS.SET_INITIAL,
        payload: { id, amount, price },
  }
}

File: src/Reducers/reducer.tsx [hunk 5]
```diff
@@ -0,0 +1,51 @@
+import { State, Action } from "../type";
```
Comment (Ishaan-flyfin) [inline src/Reducers/reducer.tsx]: change extension to .ts

File: src/Components/Product.tsx [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+import ACTIONS from "../Constants/actions";
+
+type PropType = {
+  product: productType;
+};
+
+const Product: React.FunctionComponent<PropType> = ({ product }) => {
+  const [quantity, setQuantity] = useState<number>(1);
+  const { dispatch } = useContext(CartContext);
+  const { id, img, title, price } = product;
```
Comment (Ishaan-flyfin) [inline src/Components/Product.tsx]: nice!

File: src/Components/ProductList/index.tsx [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+  if (state.productsData.length === 0) {
+    return (
+      <section className="cart">
+        <header>
+          <h2>Your Bag</h2>
+        </header>
+        <p className="empty-cart">No more Items</p>
+      </section>
+    );
```
Comment (Ishaan-flyfin) [inline src/Components/ProductList/index.tsx]: we are repeating this code
instead we can keep heading in app.js  and only return list from here

File: src/Components/ProductList/components/Product.tsx [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+      </div>
+      <div>
+        {/* Increase Button */}
+        <button
+          className="amount-btn"
+          onClick={() => {
+            setQuantity(quantity + 1);
+            dispatch({
+              type: ACTIONS.INCREASE,
+              payload: { id, amount: 1, price },
```
Comment (Ishaan-flyfin) [inline src/Components/ProductList/components/Product.tsx]: we can remove `amount: 1,`

File: src/Components/ProductList/components/Product.tsx [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+        <h4>{title}</h4>
+
+        {/* Remove Button */}
+        <button
+          className="remove-btn"
+          onClick={() => {
+            dispatch({
+              type: ACTIONS.DECREASE,
+              payload: { id, amount: quantity, price },
+            });
```
Comment (Ishaan-flyfin) [inline src/Components/ProductList/components/Product.tsx]: we can remnove this

File: src/Reducers/actions.ts [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+    total: Math.round((state.total - amount * price) * 100) / 100,
+  };
+};
+
+export const remove = (state: State, id: number) => {
+  return {
+    ...state,
+    productsData: state.productsData.filter((product) => {
+      return product.id !== id;
+    }),
```
Comment (Ishaan-flyfin) [inline src/Reducers/actions.ts]: we can update total,amount here

File: src/Reducers/actions.ts [hunk 11]
```diff
@@ -0,0 +1,30 @@
+import { State } from "../type";
+
+export const increaseAmount = (state: State, price: number) => {
+  return {
```
Comment (Ishaan-flyfin) [inline src/Reducers/actions.ts]: get product from id and fetch its price, quantity

File: src/type.ts [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+  id: number;
+  title: string;
+  price: number;
+  img: string;
+  amount: number;
```
Comment (Ishaan-flyfin) [inline src/type.ts]: we can rename it to quantity

File: src/type.ts [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+};
+
+export type State = {
+  productsData: productType[];
+  amount: number;
```
Comment (Ishaan-flyfin) [inline src/type.ts]: we can rename it to totalQuantity

**Review by Ishaan-flyfin — CHANGES_REQUESTED** [review]

File: src/App.tsx (hunk 2 above)
Comment (Snayak7643) [reply src/App.tsx]: done

File: src/Constants/InitialCartState.tsx (hunk 3 above)
Comment (Snayak7643) [reply src/Constants/InitialCartState.tsx]: done

File: src/Reducers/functions.tsx (hunk 4 above)
Comment (Snayak7643) [reply src/Reducers/functions.tsx]: done

File: src/Reducers/reducer.tsx (hunk 5 above)
Comment (Snayak7643) [reply src/Reducers/reducer.tsx]: done

File: src/type.ts (hunk 12 above)
Comment (Snayak7643) [reply src/type.ts]: done

File: src/type.ts (hunk 13 above)
Comment (Snayak7643) [reply src/type.ts]: done

File: src/Components/ProductList/components/Product.tsx (hunk 9 above)
Comment (Snayak7643) [reply src/Components/ProductList/components/Product.tsx]: done

File: src/Reducers/actions.ts (hunk 11 above)
Comment (Snayak7643) [reply src/Reducers/actions.ts]: done

File: src/Reducers/actions.ts (hunk 10 above)
Comment (Snayak7643) [reply src/Reducers/actions.ts]: done

File: src/Components/ProductList/components/Product.tsx (hunk 8 above)
Comment (Snayak7643) [reply src/Components/ProductList/components/Product.tsx]: done

File: src/Components/ProductList/index.tsx (hunk 7 above)
Comment (Snayak7643) [reply src/Components/ProductList/index.tsx]: done

File: src/Reducers/reducer.ts [hunk 14]
```diff
@@ -0,0 +1,85 @@
+import { State, Action, productType } from "../type";
+import ACTIONS from "../Constants/actionNames";
+import { initialCartState } from "../Constants/InitialCartState";
+
+const reducer = (state: State, action: Action) => {
+  //desctructuring the payload
+  const { id, quantity } = action.payload;
```
Comment (Ishaan-flyfin) [inline src/Reducers/reducer.ts]: we can avoid quantity as well

File: src/Reducers/reducer.ts [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+        sum = sum + product.price;
+      });
+
+      return {
+        ...state,
+        productsData,
+        totalQuantity: state.productsData.length,
+        totalPrice: sum,
+      };
+    }
```
Comment (Ishaan-flyfin) [inline src/Reducers/reducer.ts]: nice

File: src/Reducers/actions.ts [hunk 16]
```diff
@@ -1,30 +1,25 @@
...
+  return { type: ACTIONS.INCREASE, payload: { id } };
 };
 
-export const decreaseAmount = (state: State, amount: number, price: number) => {
-  return {
-    ...state,
-    amount: state.amount - amount,
-    total: Math.round((state.total - amount * price) * 100) / 100,
-  };
+export const decreaseAmount = (id: number, quantity: number) => {
```
Comment (Ishaan-flyfin) [inline src/Reducers/actions.ts]: quantity can be avoided here

File: src/Reducers/actions.ts [hunk 17]
```diff
@@ -1,30 +1,25 @@
...
+  return { type: ACTIONS.DECREASE, payload: { id, quantity } };
+};
+
+export const remove = (id: number, quantity: number) => {
+  return { type: ACTIONS.REMOVE, payload: { id, quantity } };
 };
 
-export const remove = (state: State, id: number) => {
+export const set_initial = (productsData: productType[]) => {
```
Comment (Ishaan-flyfin) [inline src/Reducers/actions.ts:16]: for nameing variables, function lets use camelCase

File: src/Components/ProductList/components/Product.tsx [hunk 18]
```diff
@@ -61,11 +56,11 @@ const Product: React.FunctionComponent<PropType> = ({ product }) => {
         <button
           className="amount-btn"
           onClick={() => {
+            if (quantity === 1) {
+              return dispatch(remove(id, 1));
+            }
```
Comment (Ishaan-flyfin) [inline src/Components/ProductList/components/Product.tsx]: onClick function does not have return type of dispatch
so instead `return dispatch(remove(id, 1));`
 we can do
```
dispatch(remove(id, 1));
return
```

File: src/Components/ProductList/components/Product.tsx [hunk 19]
```diff
@@ -11,11 +16,11 @@ type PropType = {
 const Product: React.FunctionComponent<PropType> = ({ product }) => {
   const [quantity, setQuantity] = useState<number>(1);
   const { dispatch } = useContext(CartContext);
-  const { id, img, title, price } = product;
+  const { id, img, title } = product;
 
   //Remove the product if quantity tends to 0
   if (quantity === 0) {
-    dispatch({ type: ACTIONS.REMOVE, payload: { id, amount: 0, price: 0 } });
+    dispatch({ type: ACTIONS.REMOVE, payload: { id, quantity: 0, price: 0 } });
```
Comment (Ishaan-flyfin) [inline src/Components/ProductList/components/Product.tsx]: do we need this now?

**Review by Ishaan-flyfin — CHANGES_REQUESTED** [review]

**Review by Ishaan-flyfin — APPROVED** [review]

**Pull request merged**